modules/substrate/src/lib.rs

```diff
@@ -371,15 +373,79 @@ impl<T: Config> bp_header_chain::HeaderChain<BridgedHeader<T>> for Module<T> {
 		PalletStorage::<T>::new().current_authority_set()
 	}
 
-	fn import_header(header: BridgedHeader<T>) -> Result<(), ()> {
-		let mut verifier = verifier::Verifier {
-			storage: PalletStorage::<T>::new(),
-		};
+	fn append_finalized_chain(headers: impl IntoIterator<Item = BridgedHeader<T>>) {
+		let mut storage = PalletStorage::<T>::new();
 
-		let _ = verifier.import_header_unchecked(header).map_err(|_| ())?;
+		for header in headers.into_iter() {
+			import_header_unchecked::<_, T>(&mut storage, header);
+		}
+	}
+}
 
-		Ok(())
+/// Import a finalized header without checking if this is true.
+///
+/// This function assumes that all the given header has already been proven to be valid and
+/// finalized. Using this assumption it will write them to storage with minimal checks. That
+/// means it's of great importance that this function *not* called with any headers whose
+/// finality has not been checked, otherwise you risk bricking your bridge.
+///
+/// One thing this function does do for you is GRANDPA authority set handoffs. However, since it
+/// does not do verification on the incoming header it will assume that the authority set change
+/// signals in the digest are well formed.
+fn import_header_unchecked<S, T>(storage: &mut S, header: BridgedHeader<T>)
+where
+	S: BridgeStorage<Header = BridgedHeader<T>>,
+	T: Config,
+{
+	// Since we want to use the existing storage infrastructure we need to indicate the fork
+	// that we're on. We will assume that since we are using the unchecked import there are no
+	// forks, and can indicate that by using the first imported header's "fork".
+	let dummy_fork_hash = <InitialHash<T>>::get();
```

Comment:
Can you elaborate why this extra storage item has to be introduced? I.e. what problem exactly does it solve?

Reply:
A wrote a little bit about it [here](https://github.com/paritytech/parity-bridges-common/pull/672#discussion_r562936774).

But the tl;dr is that I don't want to change any of the exiting storage items, so by adding this we can re-use `NextScheduledChange`.